pkg/streamingpromql/planning.go

```diff
@@ -203,6 +220,23 @@ func (p *QueryPlanner) NewQueryPlan(ctx context.Context, qs string, timeRange ty
 	return plan, err
 }
 
+func shouldWrapInDedupAndMerge(root planning.Node) bool {
+	switch node := root.(type) {
+	case *core.NumberLiteral, *core.StringLiteral:
+		return false
+	case *core.BinaryExpression:
+		if node.VectorMatching == nil {
+			return false
+		}
+	case *core.FunctionCall:
+		fn, exists := parser.Functions[node.Function.PromQLName()]
+		if exists && fn.ReturnType == parser.ValueTypeScalar {
+			return false
+		}
```

Comment:
`planning.Node` has a `ResultType()` function that might help simplify this a bit